doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,416 @@
+
+For the purposes of this document, the approach described above is sufficient to achieve a consistency level of Strict
+Serializability. However, it introduces a performance issue that we will now discuss. Consecutive writes or read/write
+cycles (read followed by write, followed by read, followed by write, etc) will increase the global timestamp in an
+unbounded fashion. If the global timestamp is larger than some object’s `upper`, and a client tries to read from that
+object using the global timestamp, then the read must be blocked until the object’s upper advances to the global
+timestamp. The larger the difference between these two values, the larger the read must be blocked for. Below I provide
+some solutions to fixing this.
+
+NOTE: This problem already exists when joining between a user table and a view over an upstream source as discussed
+in [#12198](https://github.com/MaterializeInc/materialize/issues/12198).
+
+TODO: For now all the alternatives are in the Description section (here), because I wanted to present all options before
+selecting one. Once we converge on one approach, I’ll move the rest to the Alternatives section.
+
+##### Do Nothing
+
+The “Do Nothing” approach involves doing nothing to solve this problem. We just allow reads to block if the global
+timestamp is too high.
+
+**Pros:**
+
+- Simple to implement.
+- Does not limit write throughput.
+
+**Cons:**
+
+- Allows unbounded delay on reads.
+
+##### Block Writes
+
+Before performing a write that would advance the global timestamp ahead of the current wall time, we block the write
+until the current wall time catches up to the global timestamp. As an optimization we can implement a form of group
+commit, where all writes within the same millisecond commit together at the same timestamp.
+
+**Pros:**
+
+- Prevents reads from being blocked.
+
+**Cons:**
+
+- Limits write throughput (1 write per millisecond without group commit) or limits write latency (1 millisecond per
+  write with group commit).
+
+##### Hybrid Logical Timestamps
+
+Note: There might be an actual name for this concept, but I chose the name above because it’s similar to Hybrid Logical
+Clocks.
+
+Currently, the timestamps used to in Materialize for `upper`, `since`, and the global timestamp are 64 bit unsigned
+integers that are interpreted as an epoch in milliseconds. Instead we can change the representation of these timestamps
+so that the X most significant bits are used for the epoch and the Y least significant bits are used for a counter.
+Every 1 millisecond the X bits are increased by 1 and the Y bytes are all reset to 0. Every write increments the Y bits
+by 1. This would allow us to have 2^Y writes per millisecond, without blocking any queries. Any write in
+excess of 2^Y per millisecond must be blocked until the next millisecond.
+
+Note: Increasing the size of the timestamp type would give more flexibility (for example to 128 bits).
+
+**Pros:**
```

Comment:
Probably helps us out when/if we want to make coord distributed. May allow use to use HLTs for all timestamp generation instead of the 3 distinct methods we use today.

Reply:
Added a bullet point.